### PR TITLE
 DemandManagerAssignmentList error message shows wrong object name #9320 

### LINKED ISSUE #9320: DemandManagerAssignmentList error message shows wrong object name
Issue overview
--------------
The error message for a bad meter name referenced by a DemandManagerAssignmentList will display a different object's name. 

Input object: 
```
DemandManagerAssignmentList,
    Demand Manager,          !- Name
    ElectricityPurchased:Facility,  !- Meter Name   <--- this will throw the error, should be 'Electricity:Facility' or 'ElectricityNet:Facility'
    Limit Schedule,          !- Demand Limit Schedule Name
    1.0,                     !- Demand Limit Safety Fraction
    ,                        !- Billing Period Schedule Name
    ,                        !- Peak Period Schedule Name
    15,                      !- Demand Window Length {minutes}
    SEQUENTIAL,              !- Demand Manager Priority
    DemandManager:ExteriorLights,  !- DemandManager 1 Object Type
    Ext Lights Manager,      !- DemandManager 1 Name
    DemandManager:ElectricEquipment,  !- DemandManager 2 Object Type
    Eq Mgr Stage 1,          !- DemandManager 2 Name
    DemandManager:ElectricEquipment,  !- DemandManager 3 Object Type
    Eq Mgr Stage 2,          !- DemandManager 3 Name
    DemandManager:ElectricEquipment,  !- DemandManager 4 Object Type
    Eq Mgr Stage 3,          !- DemandManager 4 Name
    DemandManager:Lights,    !- DemandManager 5 Object Type
    Lights Manager,          !- DemandManager 5 Name
    DemandManager:Thermostats,  !- DemandManager 6 Object Type
    Thermostats Manager;     !- DemandManager 6 Name
```
Error message:
```
   ** Severe  ** DemandManagerAssignmentList="COOLING PLANT" invalid value Meter Name="ELECTRICITYPURCHASED:FACILITY".
   **   ~~~   ** Only Electricity and ElectricityNet meters are currently allowed.
   **  Fatal  ** Errors found in processing input for DemandManagerAssignmentList
```
Should read DemandManagerAssignmentList="DEMAND MANAGER".
(replicated using 5ZoneAirCooledDemandLimiting.idf)

https://github.com/NREL/EnergyPlus/blob/6ec9b3bd683ea0535b823a75977dea39f4e91326/src/EnergyPlus/DemandManager.cc#L385

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA): v9.6.0
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by Nigusse**
Pull request overview
---------------------
 - Fixes #9320 
 - Fixes get input problem for DemandManagerAssignmentList object due to mix up with uninitialized variable leading to error message with wrong DemandManager object name. Regression run diffs are not expected with this change.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by Nigusse** [comment]
Below are sample error messages before and after the fix:

Before the fix:
```
   ** Severe  ** DemandManagerAssignmentList="COOLING PLANT" invalid DemandManager Name="THERMOSTATS MANAGER" not found.
   **  Fatal  ** Errors found in processing input for DemandManagerAssignmentList
```
After the fix:
```
   ** Severe  ** DemandManagerAssignmentList="DEMAND MANAGER" invalid DemandManager Name="THERMOSTATS MANAGER" not found.
   **  Fatal  ** Errors found in processing input for DemandManagerAssignmentList
```

File: src/EnergyPlus/DemandManager.cc [hunk 1]
```diff
@@ -326,64 +326,63 @@ void GetDemandManagerListInput(EnergyPlusData &state)
...
     auto &DemandMgr(state.dataDemandManager->DemandMgr);
 
-    state.dataDemandManager->NumDemandManagerList = state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, CurrentModuleObject);
+    state.dataDemandManager->NumDemandManagerList = state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, cCurrentModuleObject);
 
     if (state.dataDemandManager->NumDemandManagerList > 0) {
-        AlphArray.dimension(NumAlphas, std::string());
-        NumArray.dimension(NumNums, 0.0);
+        state.dataIPShortCut->cAlphaArgs({1, NumAlphas}) = "";
+        state.dataIPShortCut->rNumericArgs({1, NumNums}) = 0.0;
```
Comment (Nigusse) [inline src/EnergyPlus/DemandManager.cc]: Replaced local arrays `AlphArray` and `NumArray` with `state.dataIPShortCut->*`. The defect was caused due to mix up of local variable and global variable instance.

File: src/EnergyPlus/DemandManager.cc [hunk 2]
```diff
@@ -326,64 +326,63 @@ void GetDemandManagerListInput(EnergyPlusData &state)
...
 
         DemandManagerList.allocate(state.dataDemandManager->NumDemandManagerList);
 
         for (ListNum = 1; ListNum <= state.dataDemandManager->NumDemandManagerList; ++ListNum) {
 
+            auto &thisDemandMgrList = DemandManagerList(ListNum);
```
Comment (Nigusse) [inline src/EnergyPlus/DemandManager.cc]: Replaced with automatic object `thisDemandMgrList` and updated instances of the same.

File: src/EnergyPlus/DemandManager.cc [hunk 3]
```diff
@@ -393,95 +392,97 @@ void GetDemandManagerListInput(EnergyPlusData &state)
...
 
-                            if (DemandManagerList(ListNum).Manager(MgrNum) == 0) {
+                            if (thisManager == 0) {
                                 ShowSevereError(state,
-                                                CurrentModuleObject + "=\"" + state.dataIPShortCut->cAlphaArgs(1) + "\" invalid " +
-                                                    state.dataIPShortCut->cAlphaFieldNames(MgrNum * 2 + 6) + "=\"" + AlphArray(MgrNum * 2 + 6) +
-                                                    "\" not found.");
+                                                cCurrentModuleObject + "=\"" + thisDemandMgrList.Name + "\" invalid " +
+                                                    state.dataIPShortCut->cAlphaFieldNames(MgrNum * 2 + 6) + "=\"" +
+                                                    state.dataIPShortCut->cAlphaArgs(MgrNum * 2 + 6) + "\" not found.");
```
Comment (Nigusse) [inline src/EnergyPlus/DemandManager.cc]: The variable `state.dataIPShortCut->cAlphaArgs(1)` was not defined in this function and incorrectly used in the place of  `AlphArray(1)`. This line was the cause for the wrong object name. Similar incorrect uses in this function has also been fixed.

File: tst/EnergyPlus/unit/DemandResponse.unit.cc [hunk 4]
```diff
@@ -94,4 +96,91 @@ TEST_F(EnergyPlusFixture, DemandManagerGetInput)
...
+        "   **  Fatal  ** Errors found in processing input for DemandManagerAssignmentList",
+        "   ...Summary of Errors that led to program termination:",
+        "   ..... Reference severe error count=1",
+        "   ..... Last severe error=DemandManagerAssignmentList=\"DEMAND MANAGER\" invalid DemandManager Name=\"EXT LIGHTS MANAGER 1\" not found.",
+    });
+
+    EXPECT_ANY_THROW(GetDemandManagerListInput(*state));
+    EXPECT_TRUE(compare_err_stream(expected_error, true));
+}
+
```
Comment (Nigusse) [inline tst/EnergyPlus/unit/DemandResponse.unit.cc:186]: Unit test added to demonstrate the fix.

File: src/EnergyPlus/DemandManager.cc [hunk 5]
```diff
@@ -326,64 +326,63 @@ void GetDemandManagerListInput(EnergyPlusData &state)
...
 
     if (state.dataDemandManager->NumDemandManagerList > 0) {
-        AlphArray.dimension(NumAlphas, std::string());
-        NumArray.dimension(NumNums, 0.0);
+        state.dataIPShortCut->cAlphaArgs({1, NumAlphas}) = "";
+        state.dataIPShortCut->rNumericArgs({1, NumNums}) = 0.0;
 
         DemandManagerList.allocate(state.dataDemandManager->NumDemandManagerList);
 
         for (ListNum = 1; ListNum <= state.dataDemandManager->NumDemandManagerList; ++ListNum) {
```
Comment (rraustad) [inline src/EnergyPlus/DemandManager.cc]: @Nigusse there are enough changes needed to do this now and get used to best practice. Scope, declare loop, index int ListNum.

File: src/EnergyPlus/DemandManager.cc [hunk 6]
```diff
@@ -326,64 +326,63 @@ void GetDemandManagerListInput(EnergyPlusData &state)
...
+            thisDemandMgrList.Name = state.dataIPShortCut->cAlphaArgs(1);
 
-            DemandManagerList(ListNum).Meter = GetMeterIndex(state, AlphArray(2));
+            thisDemandMgrList.Meter = GetMeterIndex(state, state.dataIPShortCut->cAlphaArgs(2));
 
             if (DemandManagerList(ListNum).Meter == 0) {
-                ShowSevereError(state, "Invalid " + state.dataIPShortCut->cAlphaFieldNames(2) + '=' + AlphArray(2));
-                ShowContinueError(state, "Entered in " + CurrentModuleObject + '=' + AlphArray(1));
+                ShowSevereError(state, "Invalid " + state.dataIPShortCut->cAlphaFieldNames(2) + '=' + state.dataIPShortCut->cAlphaArgs(2));
+                ShowContinueError(state, "Entered in " + cCurrentModuleObject + '=' + thisDemandMgrList.Name);
```
Comment (rraustad) [inline src/EnergyPlus/DemandManager.cc]: format strings, e.g., ShowContinueError(state, format("Entered in {} = {}", CurrentModuleObject, thisDemandMgrList.Name)); and others like this (above and below this comment).

File: src/EnergyPlus/DemandManager.cc [hunk 7]
```diff
@@ -393,95 +392,97 @@ void GetDemandManagerListInput(EnergyPlusData &state)
...
-            DemandManagerList(ListNum).NumOfManager = int((NumAlphas - 6) / 2.0);
+            thisDemandMgrList.NumOfManager = int((NumAlphas - 6) / 2.0);
 
-            if (DemandManagerList(ListNum).NumOfManager > 0) {
-                DemandManagerList(ListNum).Manager.allocate(DemandManagerList(ListNum).NumOfManager);
+            if (thisDemandMgrList.NumOfManager > 0) {
+                thisDemandMgrList.Manager.allocate(thisDemandMgrList.NumOfManager);
 
-                for (MgrNum = 1; MgrNum <= DemandManagerList(ListNum).NumOfManager; ++MgrNum) {
+                for (MgrNum = 1; MgrNum <= thisDemandMgrList.NumOfManager; ++MgrNum) {
```
Comment (rraustad) [inline src/EnergyPlus/DemandManager.cc]: Scope, int MgrNum

**Comment by rraustad** [comment]
Develop pulled in and unit tests run locally without error.

Test file in develop, 5ZoneAirCooledDemandLimiting, no errors.

    DemandManagerAssignmentList,
      Demand Manager,          !- Name
      Electricity:Facility,    !- Meter Name

Test file in develop, 5ZoneAirCooledDemandLimiting with changed assignment list, has errors.

    DemandManagerAssignmentList,
      Demand Manager,          !- Name
      ElectricityPurchased:Facility,    !- Meter Name

    ** Severe  ** DemandManagerAssignmentList="COOLING PLANT" invalid value Meter Name="ELECTRICITYPURCHASED:FACILITY".
    **   ~~~   ** Only Electricity and ElectricityNet meters are currently allowed.
    **  Fatal  ** Errors found in processing input for DemandManagerAssignmentList

Test file in this branch corrects the warning message:

    ** Severe  ** DemandManagerAssignmentList="DEMAND MANAGER" invalid value Meter Name="ELECTRICITYPURCHASED:FACILITY".
    **   ~~~   ** Only Electricity and ElectricityNet meters are currently allowed.
    **  Fatal  ** Errors found in processing input for DemandManagerAssignmentList

**Review by rraustad — CHANGES_REQUESTED** [review]
@Nigusse this will be good to go once changes are addressed

**Comment by rraustad** [comment]
@Nigusse also pull develop and the warning will no longer appear.

File: src/EnergyPlus/DemandManager.cc (hunk 7 above)
Comment (Nigusse) [reply src/EnergyPlus/DemandManager.cc]: Done.

File: src/EnergyPlus/DemandManager.cc (hunk 5 above)
Comment (Nigusse) [reply src/EnergyPlus/DemandManager.cc]: Done.

File: src/EnergyPlus/DemandManager.cc (hunk 2 above)
Comment (amirroth) [reply src/EnergyPlus/DemandManager.cc]: I would do this as `auto &thisDemandMgrList = state.dataDemandManager->DemandManagerList(ListNum);` and delete the `auto &DemandManagerList(state.dataDemandManager->DemandManagerList);` from the top of the function.

It's not a huge deal, but I would like to move us towards using shortcut references (which are the only acceptable kinds of references IMO) in a clearer and more useful way. One aspect of that is not using references to arrays which are themselves references.

File: src/EnergyPlus/DemandManager.cc (hunk 6 above)
Comment (Nigusse) [reply src/EnergyPlus/DemandManager.cc]: Done.

**Comment by Nigusse** [comment]
@rraustad Addressed all change requests.

File: src/EnergyPlus/DemandManager.cc (hunk 2 above)
Comment (Nigusse) [reply src/EnergyPlus/DemandManager.cc]: Done. I missed the comment in my earlier push.

**Review by rraustad — APPROVED** [review]
Much cleaner now. I'll let CI process the final changes.

**Comment by rraustad** [comment]
SolarShadingTestGPU diff not related.

**Pull request merged**